contributors/devel/sig-architecture/alt-arch.md

```diff
@@ -0,0 +1,64 @@
+# Shipping alternate architectures in Kubernetes release artifacts
+
+Kubernetes by default has been amd64 oriented arch-wise. All our build and release systems originally supported just that. A while ago we started an [effort to support multiple architectures](https://github.com/kubernetes/kubernetes/issues/38067). As part of this effort, we added support in our build/release pipelines for arm, arm64, ppc64le and s390x. 
+
+The main focus was to have binaries, container images available for these architectures and for folks that are interested to be able able to take these artifacts and set up CI jobs to adequately test these platforms. Specifically to call out the ability to run conformance tests on these platforms.
+
+So in this document, let's explore with our sig-release/sig-testing hat on, what we are looking for when we talk about a new architecture
+
+# Step 1 : crawling (Build)
+
+- bazel based build infrastructure should support this architecture
+- docker based build infrastructure should support this architecture
+
+The above 2 implicitly means the following
+
+- golang should support the architecture out-of-the-box
+- All our dependencies whether vendored or run separately should support this architecture out-of-the-box
```

Comment:
Do we have a plan to demonstrate / verify that this is the case? Typically I don't think this will be an issue with most dependencies due to go's ease of cross-platform builds, but if we enforce this as a strict requirement we should have some way to measure

Reply:
We know this is not the case initially for new platforms, so it can't be a blocking strict requirement in "Step 1 : crawl (Build)".  For example there were problems with how Windows manages users and groups compared to Linux.  "Runs on platform" doesn't necessarily mean "runs equivalently" and some of this wont come out until e2e's run or maybe conformance fails.  Hopefully though that's more of an issue for extending the platform by adding OS's than when adding hardware architectures under existing OS's.  But some of this definitely shows up at compile/link time in my experiences porting various software across x86/arm/ppc/s390.  In other words "Step 1 : crawl (Build)" is iteratively improving to repeatably buildable state and that doesn't mean functional.

It's ok at the crawl level to observe in docs that the arch isn't explicitly precluded (better: is explicitly included) and then see if at compile/link time that an object is successfully generated.  This won't ensure success, but..."Step 2 : walking (Test)" follows right behind.

Reply:
> Do we have a plan to demonstrate / verify that this is the case?

building?

> Typically I don't think this will be an issue with most dependencies due to go's ease of cross-platform builds, [...]

You say that, but I recently had to fix something like this in containerd, it's easy to have use of a package like `syscall` (Go standard library, deprecated) with portability issues (in this case arm64 stopped compiling), in a project with low level OS integrations like Kubernetes it's pretty easy to run into this IMHO ..

Checking that it compiles ought to go a long way. I think the only way to automate this further is to run tests.